symlookup.1
.\" symlookup manual page
.\" Copyright © 2007-2011 Andrew Savchenko
.\"
.\" This file is part of symlookup.
.\"
.\" symlookup is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 3 as
.\" published by the Free Software Foundation
.\"
.\" symlookup is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 3 for more details.
.\"
.\" You should have received a copy of the GNU General Public License version 3
.\" along with symlookup. If not, see <http://www.gnu.org/licenses/>.
.\"
.TH symlookup 1 "28 Oct 2011" "version 0.5.2" "Symbol Lookup Utility"
.SH NAME
symlookup \- search for symbols in object files
.SH SYNOPSIS

.B symlookup
.RI "[" options "] [" symbols "]"
.br
.B symlookup
[\-h | \-\-help]
.br
.B symlookup
[\-V | \-\-version]
.br
.\" ****************************************************************
.SH DESCRIPTION

.B symlookup
is an utility intended for object symbols search in dynamic or
static symbol libraries. It looks up for libraries where given
symbols are defined and optionally can find packages holding these
libraries.

By default a search is performed in the standard system paths
recursively, looking into dynamic libraries only, and omitting
symbolic links.
But a program's behaviour may be fully customized by the command line
options.
Default system paths are determined in the same way as
.BR ld (1)
does. Refer to
.B DEFAULT PATHS
below for detailed information.

You can use extended regular expressions for symbols definition (see
.BR grep (1)
for details). Another benefit of this program is a powerful sort
engine for obtained results, see
.B SORTING
below.

Optional support for package management systems is present, it is
configurable at a compile time.
Currently
.BR rpm (8)
and
.BR portage (5)
are supported.
Fore more details refer to
.BR "PACKAGE MANAGEMENT OPTIONS" .
.\" ****************************************************************
.SH QUERY OPTIONS
.TP
.BR \-a ", " \-\-ar
Search in the
.BR ar (1)
archives altogether with shared object libraries.
.TP
.BR \-A ", " \-\-ar-only
Search in the
.BR ar (1)
archives
.IR only .
.TP
.BR \-d ", " \-\-xdev
Stay on the same physical device as a search path directory.
Useful if you don't want to descend into other file systems
mounted inside root search directories.
.TP
.BR \-F ", " \-\-filename\-repexp
Limit library file names by specified extended regular expression.
See 
.BR grep (1)
for syntax details.
Leading directory name is not affected by this filter.
Please note: extension check limitations are still active with this
option, use
.B \-X
to disable them if you really want.
.TP
.BR \-i ", " \-\-ignorecase
Ignore case in symbol names or appropriate extended regular
expressions.
.TP
.BR \-I ", " \-\-filename\-ignorecase
Ignore case in file name extended regular expression.
This option is useless without
.BR -F .
.P
.BR \-p ", "
.BI "--path " <PATH1:PATH2:...>
.RS
Specify path(s) for a library search, this will override defaults,
not extend them. Each physical file will be analysed once, even
for overlapped paths or multiple hardlinks.
.RE
.TP
.BR \-r ", " \-\-repexp
Treat given symbols as extended regular expressions likewise
.BR grep (1)
instead of search for direct match.
.TP
.BR \-s ", " \-\-follow
Follow symbolic links instead of omitting them. Be careful: often
this is just waste of time, because symbolic links usually refer
to the libraries in the same directory.
.TP
.BR \-X ", " \-\-noext
Do not perform an extension check for files in question. By default
".*\\.so\\..*" and ".*\\.so$" patterns are checked for shared
objects and ".*\\.a$" for ar archives. (If these patterns confuse
you, look at
.BR grep (1)
manual to understand them.)
.TP
.I Note:
if
.B \-a
and
.B \-A
options are specified together, the last one will take an effect.
.\" ****************************************************************
.SH OUTPUT OPTIONS
.TP
.BR \-H ", " \-\-header
Show a header when results are displayed as a table.
Only valid with
.B \-t
option.
.TP
.BR \-q ", " \-\-quiet
Be quiet. Show only fatal error messages with few exceptions. This
will also skip some non-mandatory sanity checks, thus making
program running a bit faster but less informative.
.TP
.BR \-v ", " \-\-verbose
Be more verbose than by default. This will enable some annoying
messages.
.P
.BR \-S ", "
.BI "--sort " [field,...]
.RS
Sort search results. You can arrange collation subsequence order,
see
.B SORTING
below.
.RE
.TP
.BR \-t ", " \-\-table
Print search results as a table.
.TP
.I Note:
if
.B \-q
and
.B \-v
options are specified together, the last one will take an effect.
.\" ****************************************************************
.SH PACKAGE MANAGEMENT OPTIONS

.B symlookup
has optional support for
.BR rpm (8)
and
.BR portage (5)
package management systems.
It may be enabled or disabled via configure script.
It is OK to use several package management systems at once (thought
this kind of setup is rather rare).
Of course, the following options are only valid if program was
compiled with an appropriate package management support.

.I RPM OPTIONS
.TP
.BR \-R ", " \-\-rpm
Search for rpms owning found libraries.
.TP
.BR \-\-rpm-root
Specify a path to the rpm root directory.
Please note, this is not a path to the rpm database directory, but
rather to a chroot directory holding this database; so this patch
will be prefixed to /var/lib/rpmdir or whatever is used by rpm on
your system.
.PP
.RB "An additional " rpm
search field is available for sort requests.
See
.B SORTING
below.
.PP
.B symlookup
honors rpm configuration files in the same way as
.BR rpm (8)
does, so you can attune its behaviour via rpm config files as well.

.I PORTAGE OPTIONS
.TP
.BR \-E ", " \-\-ebuild
Search for ebuilds owning found libraries.
.TP
.BR \-\-portage\-db
Specify a path to the portage database root directory.
Usually this is /var/db/pkg, but may differ if you are using prefix
setup, or chroot or similar conditions.
.PP
.RB "An additional " ebuild
search field is available for sort requests.
See
.B SORTING
below.
.TP
.I Note:
Immediate output of unsorted results is not possible for ebuild
queries because portage data base implementation is highly
inefficient for ownership queries.
All matched files will be queried at once and results will be
output in the very end of the program execution.
.\" ****************************************************************
.SH GENERAL OPTIONS
.TP
.BR \-h ", " \-\-help
Show help message, usage information and exit.
.TP
.BR \-V ", " \-\-version
Display version information and exit.
.\" ****************************************************************
.SH SORTING

.B symlookup
provides a rather powerful sort system for matches found. Thus it
may be useful not only for a particular symbols definition search,
but also for an analysis of large library sets or an entire system.

Results collation is performed sequentially according to the list
of fields provided. Fields in the list must be separated by a comma,
no spaces, tabs and other delimiters are allowed. If short sort
option
.B -S
is used the list of fields must follow the option immediately
without any spaces. If long one is used, the list must immediately
follow the '=' sign or space.

Valid field values are:
.TP
.I match
Sort results separately for an each symbol pattern provided by the
user; if neither
.BR \-i " nor " \-r
is provided this option is useless and will be ignored, because
exact match search is performed in this case. It must be the first
sort field if provided.
.TP
.I name
Sort by symbol name (case sensitive).
.TP
.I file
Sort by file name containing matched symbols.
.TP
.I ebuild
Sort by ebuild owning matched file; it is useless if
.B \-E
is unspecified and will be ignored. Option is valid only if program
is compiled with ebuild support.
.TP
.I rpm
Sort by rpm owning matched file; it is useless if
.B \-R
is unspecified and will be ignored. Option is valid only if program
is compiled with rpm support.
.P
.RE
Default sort sequence is
.IR file[,ebuild][,rpm],symbol .
This also affects
.I match
sorting. If not all fields are specified, sane defaults will be
used to complete the sequence.

Be aware of memory usage required for sorting. It is directly
proportional to a number of matched results. So if you want to
use very general regexps together with sorting, you will need
much more memory than for ordinary execution.

For this stress test:
.RS
symlookup \-arR \-\-sort=rpm,file,symbol ".*"
.RE
a peak memory usage was about 137 MB on my system with about 7000
both dynamic and static libraries installed.

If you'll ever encounter "out of memory" problem, try to
particularize search pattern or disable sort at all.
.\" ****************************************************************
.SH DEFAULT PATHS

Default search paths are detected in the same way as
.B ld (1)
utility does this, so program will check all libraries available
for an ordinary application. The following search paths are used in
order:
.TP
.B 1.
The contents of "LD_RUN_PATH" environment variable.
.TP
.B 2.
The contents of "LD_LIBRARY_PATH" environment variable.
.TP
.B 3.
The contents of "DT_RUNPATH" environment variable. If "DT_RUNPATH"
is unset "DT_RPATH" entries are checked.
.TP
.B 4.
The default system directories
.IR /lib " and " /usr/lib .
.TP
.B 5.
If the file
.I /etc/ld.so.conf
exists, the list of directories found in that file or files 
included by that file.
.PP
Note: each physical file will be analysed once, even for overlapped
paths or multiple hardlinks.
.\" ****************************************************************
.SH SECURITY CONSIDERATION

You should 
.I not
run
.B symlookup
as a root.
Though this program was carefully developed to check all possible
error conditions and to ensure NULL-terminated strings whenever
possible, it still uses intensive string I/O and there is no
guaranty buffer overflow or alike problems will never occur.
Aside from that this program use external libraries which may have
their own security issues.

For almost all cases execution as a user is sufficient. In rare
occurrence some system libraries are not world readable. If you
really want to search inside them, you should copy these libraries
somewhere and make them user-readable.
.\" ****************************************************************
.SH EXIT STATUS
.TP
.B 0
Normal exit.
.TP
.B 1
Parse error. This means that some of provided options/arguments are
incorrect. Error message provides detailed information about the
reasons of an error.
.TP
.B 2
Input/output error. Extern media is broken, perhaps bad hardware,
pipe, network connection...
.TP
.B 3
Memory allocation error. Program is running out of memory. Try to
particularize search pattern or disable results sorting. If the
last way doesn't help and you have at least some MBs of virtual
memory available please report, this may be a bug.
.TP
.B 4
Fatal libelf error. Further processing is not possible. Most
probably reason is broken libelf library, try to upgrade.
.TP
.B 5
Fatal
.BR fts (3)
function set error. Further processing is not possible. Usually
this means problems with root search directory access or
insufficient memory for directory tree structure. See error message
for details.
.\" ****************************************************************
.SH BUGS

Object files > 2GB will be skipped on 32-bit systems. But such
cases seems to be impractical.
.\" ****************************************************************
.SH AUTHOR

.B symlookup
and this manual was written by Andrew Savchenko
<bircoph@users.sourceforge.net>, feel free to mail bug reports or
feature requests to this address.
.\" ****************************************************************
.SH SEE ALSO

.BR readelf (1),
.BR objdump (1),
.BR nm (1),
.BR ar (1)
are useful utilities for examining internals of object files.
.PP
.BR elf (5)
contains an elf file format specification.
.PP
.BR find (1),
.BR locate (1),
.BR xargs (1)
are handy utilities for finding files.
.PP
.BR grep (1)
contains a description of extended regular expressions syntax.
.PP
.BR rpm (8)
can be used for examination and manipulation of rpm files.
.PP
.BR portage (5)
describes Gentoo portage packaging system.
.PP
.BR emerge (1),
.BR ebuild (1),
.BR q (1)
are powerful portage manipulation utilities.
.PP
.BR dpkg (1),
.BR dpkg-deb (1)
are the main tools to manipulate and examine deb pakges.


Changelog
symlookup v0.5.2 Fuu (ふう) Oct, 28, 2011
    * Fix default LDFLAGS for gcc-4.6.x.
    * Use reentrant strtok_r() to avoid heap corruction in elibc().
      Thanks xaionaro for bug reports.
    * Add deb package (no deb queries supported, only packaging).
    * Minor optimizations and fixes.
    * Documentation update.

symlookup v0.5.1 Pino (ぴの): Oct, 12, 2010
    * Support for older bash-3.x in configure.
    * Improve and clarify documentation.
    * Minor fixes.

symlookup v0.5.0 Sakura (桜): May, 28, 2010
    * Portage support:
      - ebuilds query;
      - support for multiple ownership;
      - fast DB queries using mmap;
      - portage DB directory is configurable.
    * Extended regular expression may be used to limit libraries
      file names.
    * Add additional verbosity level for annoying messages.
    * Man page is optimized.
    * CLI API changes: 
      --ar_only long option is replaced by --ar-only;
      -v (--version) changed to -V and -v stands for --verbose.
    * Option added to specify rpm root directory.

symlookup v0.3.5 Neko (猫): Apr, 26, 2010
    * Project name is changed to more convenient one.
    * --disable-strip is added to configure.
    * Output may be arranged in tables.
    * RPM version 5 support.
    * Support for gcc-4.4 optimizations and mfpmath sse unit.
    * Code cleanup and optimization.
    * Configure script optimization and improvement.

symbol_lookup v0.3.1 Subaru (昴): Nov, 25, 2007
    * Makefile improvements.
    * Sort for the case -sort=file[,...] fixed.
    * nftw() replaced by fts utils, thus fix open bugs.
    * Extension check improvement: discard directory name.
    * ld behavior is mimicked for default search path detection
    * The same physical file will be processed only once,
      even in case of overlapped search paths
    * Read from stdin improved.

symbol_lookup v0.3 Sachiko (幸子): Nov, 11, 2007
    * License changed to GPL v.3.
    * Optional rpm support implemented.
    * Full featured configure script created.
    * Man page created.
    * A lot of minor bugfixes and improvements.
    * Documentation update.

symbol_lookup v0.2 Misato (郷): Jun, 27, 2007
    * Regexp and case sensitiveness are implemented.
    * Configurable subsequentive sort support added.
    * Several bugs with memory allocation/dereferencing fixed.

symbol_lookup v0.1: Jun, 21, 2007
    * First draft release.
    * Basic elf and ar lookups are fully functional.



symlookup.c
/*
 *  symlookup main
 *  Copyright © 2007-2011 Andrew Savchenko
 *  Project created 04.03.2007 16:50:05 MSK (GMT+3)
 *
 *  This file is part of symlookup.
 *
 *  symlookup is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 3 as
 *  published by the Free Software Foundation
 *
 *  symlookup is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License version 3 for more details.
 *
 *  You should have received a copy of the GNU General Public License version 3
 *  along with symlookup. If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <error.h>
#include <fts.h>
#include <string.h>
#include <gelf.h>
#include <regex.h>
#include <search.h>
#include <sys/stat.h>
#include <unistd.h>

#include "symlookup.h"
#include "parser.h"
#include "safemem.h"
#include "scanelf.h"
#include "rpmutils.h"
#include "output.h"
#include "portageutils.h"

const size_t reg_error_str_len = 512;
char *reg_error_str = NULL;
//flag for found matches in the case of no sort
unsigned int matches_found = 0;

//use different compare function depend upon case sensitivity
int (*compare_func) (const char*, const char*) = strcmp;

/* structure for string array */
struct str_t
    sp       = {0, NULL}, //all search pathes (string array)
    file_arr = {0, NULL}; //matched files

/* array for matches ((char*)[4]):
   user-provided symbols
   matched files
   matched rpm
   matched ebuild */

/* match array */
struct match_arr_t match_arr = {0, NULL};

/* structure for all symbol names */
struct sym_arr symbol = {
    .size = 0,
    .match_count = NULL,
    .str    = NULL,
    .regstr = NULL,
    .match  = NULL
};

/* our options, mask flags above are used;
   unsigned int is used for speed purposes
   setup defaults */
struct opt_t opt = {
    .so   = 1,
    .ar   = 0,
    .dp   = 1,
    .ext  = 1,
    .cas  = 0,
    .tbl  = 0,
    .hdr  = 0,
#ifdef HAVE_RPM
    .rpm  = 0,
    .rpmroot = NULL,
#endif //HAVE_RPM
#ifdef HAVE_PORTAGE
    .ebuild    = 0,
    .portageDB = NULL,
#endif //HAVE_PORTAGE
    .verb = V_NORMAL,
    .re   = 0,
    .fts  = FTS_PHYSICAL,
    { /* sort */
        .cnt     = 0,
        .seq     = {0,0
#ifdef HAVE_RPM
                   ,0
#endif //HAVE_RPM
#ifdef HAVE_PORTAGE
                   ,0
#endif //HAVE_PORTAGE
                   },
        .match   = 0
    },
    .file_re = NULL
};
/* decrease M_SAVEMEM by a number of types we can save
 * in the best case*/
#if (defined(HAVE_RPM) && defined(HAVE_PORTAGE))
    unsigned int M_SAVEMEM = M_TYPES - 2;
#elif (defined(HAVE_RPM) || defined(HAVE_PORTAGE))
    unsigned int M_SAVEMEM = M_TYPES - 1;
#else
    const unsigned int M_SAVEMEM = M_TYPES;
#endif //(defined(HAVE_RPM) && defined(HAVE_PORTAGE))

/* match types */
struct mtype_t mtype = {
    .symbol = 0,
    .file = 1
};

/* free path array */
static inline void free_unused()
{
#ifdef HAVE_RPM
    //free rpm strings
    if (opt.sort.cnt)
        free(prevfile);

    /* we don't need rpm storage anymore if
       we do not use sorting */
    if (opt.rpm && !opt.sort.cnt
#ifdef HAVE_PORTAGE
            // do not free rpm array if ebuild search is engaged,
            // because immediate output is not possible in this case
            && !opt.ebuild
#endif //HAVE_PORTAGE
            ) {
        free_str(rpm_arr);
        free(rpm_arr);
    }
#endif //HAVE_RPM
    free_str(&sp);

    /* free compiled and error regexp data */
    if (opt.re || opt.file_re)
    {
        free(reg_error_str);
        if (opt.re) {
            for (unsigned int i=0; i < symbol.size; i++)
                regfree(&symbol.regstr[i]);
            free(symbol.regstr);
        }
        else
        if (opt.file_re) {
            regfree(opt.file_re);
            free(opt.file_re);
        }
    }
}

/* matched symbol manipulation */
void do_match(const unsigned int i, const char* const filename,
                                    const char* const symbolname)
{
    //don't sort => print immediately
    if (!opt.sort.cnt)
#ifdef HAVE_PORTAGE
    // due to ebuild database structure it is too expensive
    // to query ebuilds on the fly
    if (!opt.ebuild)
#endif //HAVE_PORTAGE
    {
#ifdef HAVE_RPM
        if (opt.rpm) //engage rpm support
            listrpm(filename, symbolname, NULL);
        else
#endif //HAVE_RPM
        {
            fputs(filename, stdout);
            if (!opt.tbl)
                putchar(':');
            putchar('\t');
            puts(symbolname);
        }
        matches_found = 1;
        return;
    }
    /* add new match to match_arr */
    match_arr.match = xrealloc(match_arr.match,
                               sizeof(char**) * (match_arr.count + 1));
    //allocate memory for match data
    match_arr.match[match_arr.count] = xmalloc(sizeof(char*) * M_SAVEMEM);

    static char **match;
    match = match_arr.match[match_arr.count];

    if (opt.re || opt.cas)
        match[mtype.symbol] = alloc_str(symbolname);
    else
        // In the case of exact, case insensitive match
        // we do not need to allocate new string, since
        // matched symbolname is by definition equal to
        // the user-requested symbol string.
        match[mtype.symbol] = symbol.str[i];

    /* It is possible to save some memory for multiple
       single file matches by means of small memory overhead
       for single ones and some additional CPU usage.
       Though, the most critical (at least for the first run)
       is i/o bandwidth, so I decided to do this:
       file name can be remembered only once %-) */

    if (!file_arr.size ||
        strcmp(file_arr.str[file_arr.size-1], filename)) {
        //^ reject to collate when no files are recorded
        //add new filename:
        grow_str(&file_arr, filename);

        //the same separate allocation as for filename,
        //but str_t itself
#ifdef HAVE_RPM
        if (opt.rpm) {
            rpm_arr = xrealloc(rpm_arr, sizeof(struct str_t) * file_arr.size);
            rpm_check(filename, &rpm_arr[file_arr.size-1]);
        }
#endif //HAVE_RPM
    }

    match[mtype.file] = file_arr.str[file_arr.size-1];

#ifdef HAVE_RPM
    static char *const str_rpm_nf = "<rpm not found>";
    if (opt.rpm) {
        static struct str_t *rpmname;
        rpmname = &rpm_arr[file_arr.size-1];
        //str_rpm_nf <=> no match message
        match[mtype.rpm] = (rpmname->size)? rpmname->str[0] : str_rpm_nf;

        /* unroll several rpm matches to independent match results,
           so configurable sort can be done easly */
        if (rpmname->size > 1) {
            /* add new matches to match_arr */
            match_arr.match = xrealloc(match_arr.match,
                                       sizeof(char**) * (match_arr.count + rpmname->size));

            for (unsigned int j = 1; j < rpmname->size; j++) {
                match_arr.count++;
                //allocate memory for match data
                match_arr.match[match_arr.count] = xmalloc(sizeof(char*) * M_SAVEMEM);

                match_arr.match[match_arr.count][mtype.symbol]  = match[mtype.symbol];
                match_arr.match[match_arr.count][mtype.file] = match[mtype.file];
                match_arr.match[match_arr.count][mtype.rpm]  = rpmname->str[j];
            }
        }
    }
#endif //HAVE_RPM

    /* add link to new match to symbol array;
       required only if we want to remember
       relationship with original pattern */
    if (opt.sort.match)
    {
        symbol.match[i] = xrealloc(symbol.match[i],
                                   sizeof(char***) * (symbol.match_count[i] + 1));
        symbol.match[i][symbol.match_count[i]] = match;
        symbol.match_count[i]++;

#ifdef HAVE_RPM
        //multiple rpm match leads to several formal matches
        if (opt.rpm)
        {
            static unsigned int rpmcount;
            rpmcount = rpm_arr[file_arr.size-1].size;
            if (rpmcount > 1)
            {
                symbol.match[i] = xrealloc(symbol.match[i],
                                           sizeof(char***) * (symbol.match_count[i] + rpmcount));
                //remember additional matches downwards
                for (unsigned int j = 1; j < rpmcount; j++)
                {
                    symbol.match[i][symbol.match_count[i]] = match_arr.match[match_arr.count - j];
                    symbol.match_count[i]++;
                }
            }
        }
#endif //HAVE_RPM
    }
    match_arr.count++;
}

/* comparison function for uniq_id tree */
int compare_id(register const void* const a, register const void* const b)
{
    const unsigned long long x = *(const unsigned long long* const)a;
    const unsigned long long y = *(const unsigned long long* const)b;
    return x - y;
}

/* free memory for uniq_id tree element */
void free_id(register void *const id)
{
    free(id);
}

/* Scan file tree using fts.

   We must ensure that physical files are not duplicate,
   since fts doesn't do this completely.
   Unique file id is device_id in the high 32-bit word 
   and inode_id in the low 32-bit word.
   So, unique file id is unsigned long long. 
   They are stored using balanced binary beta-tree */
static inline void fts_scan()
{
    FTS *ftsp;      //pointer to fts directory hierarchy
    FTSENT *entry;  //fts entry which depict file

    // fts_open return value isn't defined in case of errors,
    // we must check by errno 8-/
    errno=0;
    // create fts hierarchy
    ftsp = fts_open(sp.str, opt.fts, NULL);
    if (errno) {
        //save errno
        int err = errno;
        error(0,0,"fatal: invalid search path detected");
        if (opt.dp)
            error(0,0,"Check your system ld configuration at /etc/ld.so.conf and/or\n"
                      "$LD_RUN_PATH, $LD_LIBRARY_PATH, $DT_RUNPATH or $DT_RPATH environmental variables.");
        else
            error(0,0,"Check search path command line option (-p).");

        error(0,0,"Current search path array:");
        //the last element is NULL
        for (int i=0; i < sp.size-1; i++)
            error(0,0,"path[%i]='%s'",i,sp.str[i]);

        error(ERR_FTS, err, "fatal: cannot initialize file search hierarchy");
    }
    /* binary tree data */
    void *tree_root = NULL;         // pointer to id tree root pointer
    unsigned long long *file_id;    // uniq id of physical file
    struct stat *statp;             // stat structure of file in question
    void *leaf;                     // leaf in the tree
    unsigned int need_alloc = 1;    // flag for allocation memory required;

    if (opt.verb >= V_VERBOSE)
        puts("--> Iterating search tree");

    /* iterate through file hierarchy */
    while ((entry = fts_read(ftsp)))
    {   //check for error conditions
        if (opt.verb)
            switch (entry->fts_info) {
                case FTS_DC:
                    error(0,0,"warning: directory '%s' causes a cycle in the file system tree",
                          entry->fts_path);
                    continue;
                    break;
                case FTS_DNR:
                    error(0,errno,"warning: directory '%s' cannot be read", entry->fts_path);
                    continue;
                    break;
                case FTS_ERR:
                    error(0,errno,"warning: error accessing file '%s'", entry->fts_path);
                    continue;
                    break;
                case FTS_NS:
                    error(0,errno,"warning: cannot stat file '%s'", entry->fts_path);
                    continue;
                    break;
                case FTS_SLNONE:
                    error(0,0,"warning: file '%s' is a stale symbolic link", entry->fts_path);
                    continue;
                    break;
            }
        //process only regular files
        if (entry->fts_info == FTS_F)
        {
            //create unique id
            statp = entry->fts_statp;
            //allocate memory only if previous id was put in the tree
            if (need_alloc)
                file_id = xmalloc(sizeof(unsigned long long));
            *file_id = (unsigned long long)(statp->st_dev) << 32 | statp->st_ino;

            //check new file
            need_alloc=1;
            if (!(leaf = tsearch(file_id, &tree_root, compare_id))) {
                if (opt.verb)
                    error(0,errno,"error: not enough memory for file id tree, "
                                  "search results may be duplicated");
                need_alloc=0;
            } else
            //skip already checked file
            if (*(unsigned long long**)leaf != file_id) {
                need_alloc=0;
                continue;
            }
            checkfile(entry->fts_accpath, entry->fts_path, entry->fts_name);
        }
    }
    // fts_read() sets errno to 0 explicitly if all was ok
    if (errno && opt.verb)
        error(0, errno, "warning: fts hierarchy scan was ended abnormally,\n"
                        "search results may be incomplete");
    //free field_id pointer if it isn't in the tree
    if (!need_alloc)
        free(file_id);
    //free search tree
    tdestroy(tree_root, free_id);

    if (fts_close(ftsp) == -1 && opt.verb)
        error(0, errno, "warning: can't close fts file hierarchy stream\n"
                        "and restore working directory");
}

int main(const int argc, char *const argv[])
{
    /* parse args & preinit some vars */
    parse(argc, argv);

    /* accomplish symbol base init only if sort requested */
    if (opt.sort.match) {
        symbol.match = xcalloc (symbol.size, sizeof(char****));
        symbol.match_count = xcalloc (symbol.size, sizeof(int));
    }

    /* init libelf */
    if (elf_version(EV_CURRENT) == EV_NONE)
        error(ERR_ELF, elf_errno(), "fatal: cannot initialize libelf");

    /* prepare output */
    init_output();

    /* scan file hierarchy */
    fts_scan();

    /* free unneeded memory */
    free_unused();

#ifdef HAVE_PORTAGE
    /* search for ebuilds owning files in questions */
    if (opt.ebuild)
        find_ebuilds(&file_arr);
#endif //HAVE_PORTAGE

    /* sort if required and output results */
    if (opt.sort.cnt)
        sort_output();
#ifdef HAVE_PORTAGE
    else // special case for unsorted ebuild output as it can't be done in do_match()
    if (opt.ebuild)
        ebuild_unsorted_output();
#endif //HAVE_PORTAGE
    else
    if (opt.verb && !matches_found)
        puts(str_not_found);

#ifdef HAVE_RPM
    /* uninit rpm */
    if (opt.rpm)
        rpmuninit();
#endif //HAVE_RPM
}

